src/urllib3/_collections.py

```diff
@@ -45,54 +100,67 @@ class RecentlyUsedContainer(MutableMapping):
         ``dispose_func(value)`` is called.  Callback which will get called
     """
 
-    ContainerCls = OrderedDict
+    _container: "OrderedDict[_KT, _VT]"
+    _maxsize: int
+    dispose_func: Optional[Callable[[_VT], None]]
+    lock: RLock
 
-    def __init__(self, maxsize=10, dispose_func=None):
+    def __init__(
+        self, maxsize: int = 10, dispose_func: Optional[Callable[[_VT], None]] = None
+    ) -> None:
+        super().__init__()
         self._maxsize = maxsize
         self.dispose_func = dispose_func
-
-        self._container = self.ContainerCls()
+        self._container = OrderedDict()
         self.lock = RLock()
 
-    def __getitem__(self, key):
+    def __getitem__(self, key: _KT) -> _VT:
         # Re-insert the item, moving it to the end of the eviction line.
         with self.lock:
             item = self._container.pop(key)
             self._container[key] = item
             return item
 
-    def __setitem__(self, key, value):
-        evicted_value = _Null
+    def __setitem__(self, key: _KT, value: _VT) -> None:
+        evicted_value = None
         with self.lock:
             # Possibly evict the existing value of 'key'
-            evicted_value = self._container.get(key, _Null)
+            try:
+                # If the key exists, we'll overwrite it, which won't change the
+                # size of the pool. Because accessing a key should move it to
+                # the end of the eviction line, we pop it out first.
+                evicted_value = self._container.pop(key)
+            except KeyError:
+                if len(self._container) >= self._maxsize:
+                    # If we didn't evict an existing value, and we've hit our maximum
+                    # size, then we have to evict the least recently used item from
+                    # the beginning of the container.
+                    _, evicted_value = self._container.popitem(last=False)
+
+            # Finally, insert the new value.
             self._container[key] = value
 
-            # If we didn't evict an existing value, we might have to evict the
-            # least recently used item from the beginning of the container.
-            if len(self._container) > self._maxsize:
-                _key, evicted_value = self._container.popitem(last=False)
-
-        if self.dispose_func and evicted_value is not _Null:
+        # Release the lock on the pool, and dispose of the evicted value.
+        if evicted_value is not None and self.dispose_func:
```

Comment:
Isn't it now possible that an evicted `None` value won't have `dispose_func` called? Maybe we need a flag that is set if we actually evict anything. Not sure if this actually matters, but I do know `None` is inserted into ConnectionPools.

Reply:
That's a good catch; this would not call the eviction function for some `evicted_value` if `_VT` is a supertype of `None`. Will play with some things.